src/audio/pcm_converter/pcm_converter_hifi3.c

```diff
@@ -267,79 +196,53 @@ static int pcm_convert_s16_to_s32(const struct audio_stream __sparse_cache *sour
 				  uint32_t ioffset, struct audio_stream __sparse_cache *sink,
 				  uint32_t ooffset, uint32_t samples)
 {
-	ae_int16 *in = audio_stream_read_frag(source, ioffset,
-					      sizeof(int16_t));
-	ae_int32 *out = audio_stream_write_frag(sink, ooffset,
-						sizeof(int32_t));
+	int16_t *src = source->r_ptr;
+	int32_t *dst = sink->w_ptr;
 	ae_int16x4 sample = AE_ZERO16();
-	ae_valign align_out = AE_ZALIGN64();
-	ae_int16x4 *in16x4;
-	ae_int32x2 *out32x2;
-	int i = 0;
+	int nmax, i, n, m, left;
+	ae_valign inu = AE_ZALIGN64();
+	ae_valign outu = AE_ZALIGN64();
+	int left_samples = samples;
+	ae_int16x4 *in = audio_stream_wrap(source, src + ioffset);
+	ae_int32x2 *out = audio_stream_wrap(sink, dst + ooffset);
 
 	/* nothing to do */
 	if (!samples)
 		return samples;
 
-	/* required alignment for AE_L16X4_XC */
-	while (!IS_ALIGNED((uintptr_t)in, 8)) {
-		/* set source as circular buffer */
-		pcm_converter_setup_circular(source);
-
-		/* load one 16 bit sample */
-		AE_L16_XC(sample, in, sizeof(ae_int16));
-
-		/* set sink as circular buffer */
-		pcm_converter_setup_circular(sink);
-
-		/* store one 32 bit sample */
-		AE_S32_L_XC(AE_CVT32X2F16_32(sample), out, sizeof(ae_int32));
-
-		if (++i == samples)
-			return samples;
-	}
-
-	in16x4 = (ae_int16x4 *)in;
-	out32x2 = (ae_int32x2 *)out;
-
-	/* main loop processes 4 samples at a time */
-	while (samples >= 3 && i < samples - 3) {
-		/* set source as circular buffer */
-		pcm_converter_setup_circular(source);
-
-		/* load four 16 bit samples */
-		AE_L16X4_XC(sample, in16x4, sizeof(ae_int16x4));
-
-		/* set sink as circular buffer */
-		pcm_converter_setup_circular(sink);
-
-		/* store two 32 bit samples */
-		AE_SA32X2_IC(AE_CVT32X2F16_32(sample), align_out, out32x2);
-		AE_SA32X2_IC(AE_CVT32X2F16_10(sample), align_out, out32x2);
-
-		i += 4;
+	while (left_samples > 0) {
```

Comment:
we could micro-optimise this a bit further: we checked that `samples != 0` above already, so, this loop must run at least once. So, we can use a `do {} while (left_samples > 0)` loop instead here. The advantage would be, that with a `while` loop (as long as the compiler isn't smart enough, which it likely isn't) the CPU has to perform a jump at the end of the loop back to the beginning to check the condition, then it will in most cases find, that it doesn't need to wrap and it will jump back. So, it will perform two jumps. With a `do {} while ()` loop it will check the condition at the bottom of the loop, and in those cases without wrapping won't jump at all and will just proceed linearly with the code after the loop.

Reply:
since has removed the samples != 0 check, so I'd like to keep the while loop.